library/psa_crypto.c

```diff
@@ -3975,9 +3975,83 @@ psa_status_t psa_cipher_update( psa_cipher_operation_t *operation,
         goto exit;
     }
 
-    ret = mbedtls_cipher_update( &operation->ctx.cipher, input,
-                                 input_length, output, output_length );
-    status = mbedtls_to_psa_error( ret );
+    if( operation->alg == PSA_ALG_ECB_NO_PADDING )
+    {
+        /* mbedtls_cipher_update has an API inconsistency: it will only
+        * process a single block at a time in ECB mode. Abstract away that
+        * inconsistency here to match the PSA API behaviour. */
+        *output_length = 0;
+
+        if( input_length == 0 )
+        {
+            status = PSA_SUCCESS;
+            goto exit;
+        }
+
+        if( expected_output_size > 0 )
+        {
+            size_t ctx_bytes = operation->ctx.cipher.unprocessed_len;
+            if( ctx_bytes > 0 )
+            {
+                /* Fill up to block size and run the block */
+                size_t bytes_to_copy = operation->block_size - ctx_bytes;
+                memcpy( &( operation->ctx.cipher.unprocessed_data[ctx_bytes] ),
+                        input, bytes_to_copy );
+                input_length -= bytes_to_copy;
+                input += bytes_to_copy;
+                operation->ctx.cipher.unprocessed_len = 0;
+
+                status = mbedtls_to_psa_error(
+                    mbedtls_cipher_update( &operation->ctx.cipher,
+                                           operation->ctx.cipher.unprocessed_data,
+                                           operation->block_size,
+                                           output, &internal_output_length ) );
+
+                if( status != PSA_SUCCESS )
+                    goto exit;
+
+                output += internal_output_length;
+                output_size -= internal_output_length;
+                *output_length += internal_output_length;
+            }
+
+            size_t blocks = input_length / operation->block_size;
+            for( ; blocks > 0; blocks-- )
```

Comment:
Please rewrite the loop to use something like
```
for( size_t n = input_length; n >= operation->block_size; n -= operation->block_size )
```
avoiding division by the block size.

We avoid division with a non-constant in symmetric cryptography code because on some platforms without hardware division, it requires pulling in a library function. This is admittedly not well documented. There is a configuration option `MBEDTLS_NO_UDBL_DIVISION` which informs the choice of word size in the bignum module, and no part of the crypto library should use division on a larger type (apart from division by a constant). A platform could have a 64-bit `size_t` and a 32-bit `mbedtls_mpi_uint` and no 64-bit hardware division, although I'd expect that on most such platforms linking a software division function wouldn't be a problem. We have no such platform on our CI (we test the effectiveness of `MBEDTLS_NO_UDBL_DIVISION` on a 32-bit arm build, with a 32-bit `size_t`). I don't think it's worth taking the risk, even if it's only in a somewhat “exotic” platform. Here the code can easily be rewritten.

Reply:
Done.

Reply:
Looking further, I see that there was already a division by the block size in `psa_crypto.c`, and [a few other similar cases in the code base](https://github.com/ARMmbed/mbedtls/issues/3654). I don't think this was intended, but some have been around for years, so I don't think they bother anyone. I know that 64-bit division on 32-bit platforms is an actual problem that some users of Mbed TLS have faced, but division by `size_t` apparently isn't.